text/entities/0264-resource-and-entities.md

```diff
@@ -0,0 +1,806 @@
+  - Resource detectors otherwise follow existing merge semantics.
+    - The Specification merge rules will be updated to account for violations prevalent in ALL implementation of resource detection.
+    - Specifically: This means the [rules around merging Resource across schema-url will be dropped](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge).  Instead only conflicting attributes will be dropped.
+    - SchemaURL on Resource will be deprecated with entity-specific schema-url replacing it. SDKs will only fill out SchemaURL on Resource when SchemaURL matches across all entities discovered. Additionally, only existing stable resource attributes can be used in Resource SchemaURL in stable OpenTelemetry components (Specifially `service.*` and `sdk.*` are the only stabilized resource convnetions). Given prevalent concerns of implementations around Resource merge specification, we suspect impact of this deprecation to be minimal, and existing usage was within the "experimental" phase of semantic conventions.
+  - An OOTB ["Env Variable Entity Detector"](#environment-variable-detector) will be specified and provided vs. requiring SDK wide ENV variables for resource detection.
+- *Additionally, Resource Coordinator would be responsible for understanding Entity lifecycle events, for Entities whose lifetimes do not match or exceed the SDK's own lifetime (e.g. browser session).*
+
+#### Entity Detector
+
+The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK.  For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod.   SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.
+
+An Entity Detector would have an API similar to:
+
+```rust
+trait EntityDetector
+  pub fn detect_entities(...) -> Result<Vec<Entity>, EntityDetectionError>>
+```
+
+Where `Result` is the equivalent of error channel in the language of choice (e.g. in Go this would be `entities, err := e.detectEntities()`).
+
+An Entity Detector MUST NOT provide two entities of the same entity type (e.g. two `host` or two `service` entities).
+
+#### Entity Merging and Resource
+
+The most important aspect of this design is how Entities will be merged to construct a Resource.
+
+We provide a simple algorithm for this behavior:
+
+- Construct a set of detected entities, `E`
+  - All entity detectors are sorted by priority (highest first)
+  - For each entity detector `D`, detect entities
+    - For each entity detected, `d'`
+      - If an entity `e'` exists in `E` with same entity type as `d'`, do one of the following:
+        - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:
```

Comment:
Please clarify this sentence:

> - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:

Who's "entity" in _If the entity..._? I can interpret this as:

``` python
if e.identity == e.schema_url:
	merge_first_into_second(d, e)
```

And we probably don't want to compare different attributes of the same entity, right?

Do entities have a `schema_url` attribute? From [this](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#data-model) I was under the impression an entity only attributes were `type`, `id` and `attributes`.

Also, isn't `identity` a `map<string, attribute>`? How can we compare it to `schema_url` which is a string?

Reply:
> Do entities have a schema_url attribute? From [this](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#data-model) I was under the impression an entity only attributes were type, id and attributes.

Yes entities have a schema_url.

> Also, isn't identity a map<string, attribute>? How can we compare it to schema_url which is a string?

Identity is a `KeyValueList`.  You compare identity *and* you compare schema_url.  I.e. you check if:

`lhs.schema_url == rhs.schema_url && lhs.identifying_attributes == rhs.identifying_attributes`

Reply:
I think it is somewhat confusing that schema_*url* is treated as an opaque value consistently here. It seems that there is a silent assumption that it would be a bad idea to actually fetch the content of the URL and buffer it locally.

Reply:
Yes: This was explicitly talked about but didn't make it into the proposal.  We need to be able to execute our algorithm in environments where there is no access to public internet, and users may not have silently provided internal access to these URLs.

Therefore while we can *allow* it to be fetched and buffered (and we plan to use it in such a way) the core algorithm must work without this.

Reply:
I added a section to call out this assumption - see the "Why don't we download schema url contents?" section

Reply:
@jsuereth thanks; Imho that section resolves the issue in the current state of the elaboration.